### ISSUE TITLE
Tiled Crashes When Opening A File

### ISSUE THREAD

**Issue by secv** [bjorn/tiled#2675]
Program may crash each time a single file is opened. The file may open normally as well. The crashing file can be opened and operates normally when selected as part of a world (selecting the faded out map in the editor window). The error message is included below:

> Problem signature:
>   Problem Event Name:	APPCRASH
>   Application Name:	tiled.exe
>   Application Version:	1.0.0.0
>   Application Timestamp:	5dcbd3ce
>   Fault Module Name:	StackHash_6608
>   Fault Module Version:	0.0.0.0
>   Fault Module Timestamp:	00000000
>   Exception Code:	c0000005
>   Exception Offset:	0000000004e5fd3c
>   OS Version:	6.1.7601.2.1.0.768.3
>   Locale ID:	1033
>   Additional Information 1:	6608
>   Additional Information 2:	6608dae5bcac251cbc87eead97ed4e3a
>   Additional Information 3:	9add
>   Additional Information 4:	9add1cf46553aa5279cdf865cc8dda04

**Comment by bjorn** [comment]
Hmm, this error message unfortunately contains no relevant information at all. Would you be able to give me access to the files causing this crash? I haven't generally noticed any crash simply on opening of a file.

**Comment by bjorn** [comment]
Also, I assume you're reporting this against Tiled 1.3.0? Did you use previous versions of Tiled and were they also affected by this issue?

**Comment by secv** [comment]
Yes, I am using 1.3.0. Previous versions did crash, but not often or in the same manner (repeatedly crashing on the same file, then not crashing). I have to see if I can post the files or in what ways I can send them as they are part of a private project.

I do have the following warnings reporting in the Tiled Issues panel. Maybe they are related to the crashes:

![Untitled-1](https://user-images.githubusercontent.com/4642505/69241794-880b3c80-0b6d-11ea-9ff5-459453960aa0.jpg)

**Comment by bjorn** [comment]
> Yes, I am using 1.3.0. Previous versions did crash, but not often or in the same manner (repeatedly crashing on the same file, then not crashing). I have to see if I can post the files or in what ways I can send them as they are part of a private project.

If the private project is on GitHub you could temporarily add me as a collaborator. You can also reach me privately at bjorn@lindeijer.nl.

> I do have the following warnings reporting in the Tiled Issues panel. Maybe they are related to the crashes:

Unfortunately those warnings have nothing to do with the crash. The libpng warning refers to the images in your project, which apparently contain an incorrect sRGB profile. In previous versions the warning was simply not visible, but like this it's also not very useful. I should look into making it show the name of the image, but I think that'll get hacky.

The "TLS initialization failed" warning is unfortunately due to the Windows 1.3.0 release shipping with the wrong OpenSSL DLLs, which means new version notification isn't working. It is fixed in the 1.3.1 release I'm making today.

**Comment by tomcashman** [comment]
I'll send you over our map files now :)

**Comment by bjorn** [comment]
I fixed the issue with the wrong world getting used, so now you can keep all your worlds loaded. I've triggered a snapshot build so you can test whether it also avoids the crash.

**Comment by tomcashman** [comment]
I was getting the issue again this morning but not happening on the snapshot. However, it does seem to occur randomly so we'll report back over next few days :)

I was thinking it could be world related if it was loading all 4 of our worlds and Tiled was running OOM. Might be worth looking into adding crash logs or something like [crashpad/sentry](https://docs.sentry.io/platforms/native/crashpad/).

**Comment by bjorn** [comment]
> I was getting the issue again this morning but not happening on the snapshot.

Alright, just make sure you've upgraded to snapshot version 2019.11.26.

> I was thinking it could be related to world loading if it was loading all 4 of our worlds and Tiled was running OOM. Might be worth looking into adding crash logs or something like [crashpad/sentry](https://docs.sentry.io/platforms/native/crashpad/).

I assume you're not using the 32-bit WinXP version of Tiled. For the 64-bit version of Tiled to run out of memory would be very unlikely. With maps from all 4 worlds loaded Tiled consumes 0.5 GB resident / 2.7 GB virtual memory for me on Linux.

Integrating a crash handler would certainly be helpful! I see sentry.io handles the server part of this, which should hugely simplify the process. I'll have to look into setting that up. I think last time I heard about sentry it may not have supported native code yet. Thanks for the link!

**Comment by bjorn** [comment]
@secv @tomcashman If you haven't run into the crash for the past two weeks using the snapshot version then I'd suggest we close this issue.

I've looked into enabling Sentry for a few hours btw, but have not been able to actually get it to send any crash reports (just sending a test event worked). I've only tested the Linux version so far though, so I'll still look into enabling it on Windows at some point. See issue #2681 for that.

**Comment by tomcashman** [comment]
I'm actually encountering it today. Is there any way to get logs or debug?

**Comment by bjorn** [comment]
> I'm actually encountering it today. Is there any way to get logs or debug?

At the moment only if you'd run a debug build of Tiled and reproduce the issue in a debugger, for which you'd currently need to compile Tiled locally since the releases don't have debug information.

Anyway, thanks for letting me know! Unless you have someone on your team who can debug this I suppose that increases the priority of integrating a crash handler in the Windows builds.

**Comment by tomcashman** [comment]
@bjorn has there been any progress on crash handling/logging? Still getting this issue regularly on latest version.

**Comment by bjorn** [comment]
I have looked into it back in December and looking into it again the past two days. So far, it does not seem like I'm getting anywhere unfortunately.

I'm looking at integrating [Sentry Native](https://docs.sentry.io/platforms/native/), but running into one weird problem after the other. On Linux I got it to compile and run, but when I crash it, it just hangs without any further feedback. On Windows I so far can't get it to compile with MinGW. After finally getting it to find the compiler at all, cmake seems hell bent on passing Visual Studio compiler options when it wants to build zlib. Maybe I'll give it a try on macOS next...

In any case, I'm sorry for the lack of progress on this. I'd love to have a crash handler integrated but it will take a lot of trial and error. Maybe it would be easier for you to run a self-compiled version of Tiled in a debugger? I could help you with it if we can do a video call with screen sharing for example.
